### PR TITLE
[fundamental] Recording support metrics

### PR DESCRIPTION
**Pull request by crazygao**
# Description

How to support metrics:
Old injection inject mock_tool, and all logic below mock_tool is dropped. especially for metric collection
Good point. mock_tool could also record items with no llm calls.

New way:
add a new layer at openai_inject.
In this way, the logic below mock_tool is not affected. For openai calls, it is easily collected.
Bad point, it doesn't know other info besides openai call.

So the new recording is a combination of these two.
Inject openai calls for llm nodes. + Inject tool for python nodes.

Also upload token numbers in live mode. Here is an example pipeline
https://github.com/microsoft/promptflow/actions/runs/7705912004

And we can retrieve the uploaded artifact.

image

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/conftest.py [hunk 1]
```diff
@@ -20,7 +20,8 @@
 from dotenv import load_dotenv
 from filelock import FileLock
 from pytest_mock import MockerFixture
-from sdk_cli_azure_test.recording_utilities import SanitizedValues, is_replay
+from sdk_cli_azure_test.recording_utilities import SanitizedValues, is_record, is_replay
```
Comment (wangchao1230) [inline src/promptflow/tests/conftest.py:23]: in PR description pls include what's the main change to enable metrics support

Comment (crazygao) [reply src/promptflow/tests/conftest.py:23]: Okay, this is quite a long story...

Comment (crazygao) [reply src/promptflow/tests/conftest.py:23]: I'll make it after next iteration.

File: src/promptflow/tests/sdk_cli_test/conftest.py [hunk 2]
```diff
@@ -8,9 +8,11 @@
 import pytest
 from mock import mock
 from pytest_mock import MockerFixture
+from sdk_cli_azure_test.recording_utilities import is_live, is_record, is_replay
```
Comment (zhengfeiwang) [inline src/promptflow/tests/sdk_cli_test/conftest.py:11]: can we make this import better? kind of wired to import a azure fixture in a local test conft...

File: src/promptflow/tests/sdk_cli_test/recording_utilities/openai_inject_recording.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            return call_func(f, args, kwargs)
+
+    return wrapper
+
+
+def inject_async_with_recording(f):
+    wrapper_fun = inject_operation_headers(
+        (inject_function_async(["api_key", "headers", "extra_headers"])(inject_recording(f)))
```
Comment (zhengfeiwang) [inline src/promptflow/tests/sdk_cli_test/recording_utilities/openai_inject_recording.py:27]: nit. add parameter name here, though these lines will become longer, but it's really hard to understand this line without comment, but add parameter name should help a lot.

Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_test/recording_utilities/openai_inject_recording.py:27]: something like below:

```python
(inject_function_async(args_to_ignore=["api_key", "headers", "extra_headers"])(inject_recording(f)))
```

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/tests/sdk_cli_test/conftest.py (hunk 2 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/conftest.py:11]: Okay

**Review by zhengfeiwang — DISMISSED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by chw-microsoft — APPROVED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Pull request merged**